lightning-persister/src/lib.rs

```diff
@@ -28,9 +27,38 @@ use lightning::ln::channelmanager::ChannelManager;
 use lightning::util::logger::Logger;
 use lightning::util::ser::{ReadableArgs, Writeable};
 use std::fs;
-use std::io::{Cursor, Error};
+use std::io::{Cursor, Error, Write};
 use std::ops::Deref;
-use std::path::{Path, PathBuf};
+use std::path::{Path, PathBuf, MAIN_SEPARATOR};
+
+
+/// Trait that handles persisting a [`ChannelManager`] and [`NetworkGraph`] to disk.
+pub trait Persister
+{
+	/// Persist the given [`ChannelManager`] to disk, returning an error if persistence failed
+	/// (which will cause the BackgroundProcessor which called this method to exit).
+	fn persist_manager<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref>(&self, channel_manager: &ChannelManager<Signer, M, T, K, F, L>) -> Result<(), std::io::Error> where
+	M::Target: 'static + chain::Watch<Signer>,
+	T::Target: 'static + BroadcasterInterface,
+	K::Target: 'static + KeysInterface<Signer = Signer>,
+	F::Target: 'static + FeeEstimator,
+	L::Target: 'static + Logger;
+
+	/// Persist the given [`NetworkGraph`] to disk, returning an error if persistence failed.
+	fn persist_graph(&self, network_graph: &NetworkGraph) -> Result<(), std::io::Error>;
+}
+
+/// Trait for a key-value store for persisting some writeable object at some key
+pub trait KVStorePersister<W: Writeable> {
+	/// Persist the given writeable using the provided key
+	fn persist(&self, key: String, object: &W) -> std::io::Result<()>;
+}
+
+impl<W: Writeable> KVStorePersister<W> for FilesystemPersister {
+	fn persist(&self, key: String, object: &W) -> std::io::Result<()> {
+		util::write_to_file(format!("{}{}{}", self.path_to_channel_data, MAIN_SEPARATOR, key), object)
```

Comment:
Oh, right, I forgot we build the key first and then want to push a prefix onto it...means we can't avoid a second allocation here.